src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
+    m_file_lists("f", "files", "read newline separated list of files to check (see --examine) from the specified path", false, "PATH", m_parser),
+    m_root_path("r", "root", "check files relative to the given root directory", false, "", "PATH", m_parser),
+    m_config_root_path("", "config-root", "lookup configuration files relative to the given root directory", false, "", "PATH", m_parser),
+    m_input_args("args", "in --system mode a list of paths to check, otherwise a list of profiles to parse", false, "PATH", m_parser),
+    m_euid(geteuid())
 {
-    struct perm *next;
-    char *file;
-    char *owner;
-    char *group;
-    mode_t mode;
-    cap_t caps;
-};
-
-struct perm *permlist;
-char **checklist;
-size_t nchecklist;
-uid_t euid;
-char *root;
-size_t rootl;
-size_t nlevel;
-const char** level;
-int do_set = -1;
-int default_set = 1;
-int have_fscaps = -1;
-char** permfiles = NULL;
-size_t npermfiles = 0;
-char* force_level;
-std::string config_root;
-
-struct perm*
-add_permlist(char *file, const char *p_owner, const char *p_group, mode_t mode)
+}
+
+bool Chkstat::validateArguments()
 {
-    struct perm *ec, **epp;
+    // exits on error/usage
+    m_parser.parse(m_argc, m_argv);
 
-    char *owner = strdup(p_owner);
-    char *group = strdup(p_group);
-    if (rootl)
+    // check all parameters and only then return to provide full diagnostic to
+    // the user, not just bit by bit complaining.
+    bool ret = true;
+
+    const auto xor_args = {
+        std::make_pair(&m_system_mode, static_cast<TCLAP::SwitchArg*>(&m_apply_changes)),
+        {&m_force_fscaps, &m_disable_fscaps},
+        {&m_apply_changes, &m_only_warn}
+    };
+
+    for (const auto &args: xor_args)
     {
-        char *nfile;
-        nfile = (char*)malloc(strlen(file) + rootl + (*file != '/' ? 2 : 1));
-        if (nfile)
+        auto &arg1 = *args.first;
+        auto &arg2 = *args.second;
+
+        if (arg1.isSet() && arg2.isSet())
         {
-            strcpy(nfile, root);
-            if (*file != '/')
-                strcat(nfile, "/");
-            strcat(nfile, file);
+            std::cerr << arg1.getName() << " and " << arg2.getName()
+                << " cannot be set at the same time\n";
+            ret = false;
         }
-        file = nfile;
-    }
-    else
-    {
-        file = strdup(file);
     }
 
-    if (!owner || !group || !file)
+    if (!m_system_mode.isSet() && m_input_args.getValue().empty())
     {
-        perror("permlist entry alloc");
-        exit(1);
+        std::cerr << "one or more permission file paths to use are required\n";
+        ret = false;
     }
 
-    for (epp = &permlist; (ec = *epp) != 0; )
+    for (const auto arg: { &m_root_path, &m_config_root_path })
     {
-        if (!strcmp(ec->file, file))
+        if (!arg->isSet())
+            continue;
+
+        auto &path = arg->getValue();
+
+        if (path.empty() || path[0] != '/')
```

Comment:
Yeah it was this way before. When supporting relative paths we'd need to take the CWD into account ... either trust that the CWD is always correct and unaltered or we'd need to explicitly resolve relative paths to absolute ones. It would be more like a new feature that I don't necessarily want in the refactoring.

Reply:
OK, I still don't see the value of this check (it seems like one of these ill-gotten 'security' checks that were never thought through) - but if it was like this before, we can keep this as-is.